File: src/main/java/org/jukeboxmc/network/handler/LoginHandler.java
package org.jukeboxmc.network.handler;

import org.cloudburstmc.protocol.bedrock.packet.LoginPacket;
import org.cloudburstmc.protocol.bedrock.packet.PlayStatusPacket;
import org.cloudburstmc.protocol.bedrock.packet.ResourcePacksInfoPacket;
import org.jukeboxmc.Server;
import org.jukeboxmc.event.player.PlayerLoginEvent;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.player.data.LoginData;
import org.jukeboxmc.resourcepack.ResourcePack;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class LoginHandler implements PacketHandler<LoginPacket> {

    @Override
    public void handle( LoginPacket packet, Server server, Player player ) {
        player.getPlayerConnection().setLoginData( new LoginData( packet ) );

        PlayerLoginEvent playerLoginEvent = new PlayerLoginEvent( player );

        if ( !player.getPlayerConnection().getLoginData().isXboxAuthenticated() ) {
            playerLoginEvent.setResult( PlayerLoginEvent.Result.XBOX_AUTHENTICATED );
            playerLoginEvent.setKickReason( "You must be logged in with your xbox account." );
        }

        if ( Server.getInstance().getOnlinePlayers().size() >= Server.getInstance().getMaxPlayers() ) {
            playerLoginEvent.setResult( PlayerLoginEvent.Result.SERVER_FULL );
            playerLoginEvent.setKickReason( "Server is full." );
        }
        if ( server.hasWhitelist() ) {
            if ( !server.isPlayerOnWhitelist( player.getName() ) ) {
                playerLoginEvent.setResult( PlayerLoginEvent.Result.WHITELIST );
                playerLoginEvent.setKickReason( "You are not on the whitelist" );
            }
        }
        Server.getInstance().getPluginManager().callEvent( playerLoginEvent );

        if ( playerLoginEvent.getResult() != PlayerLoginEvent.Result.ALLOWED ) {
            player.kick( playerLoginEvent.getKickReason() );
            return;
        }

        player.getPlayerConnection().sendPlayStatus( PlayStatusPacket.Status.LOGIN_SUCCESS );

        ResourcePacksInfoPacket resourcePacksInfoPacket = new ResourcePacksInfoPacket();
        for ( ResourcePack resourcePack : server.getResourcePackManager().retrieveResourcePacks() ) {
            resourcePacksInfoPacket.getBehaviorPackInfos().add(
                    new ResourcePacksInfoPacket.Entry(
                            resourcePack.getUuid().toString(),
                            resourcePack.getVersion(),
                            resourcePack.getSize(),
                            "",
                            "",
                            resourcePack.getUuid().toString(),
                            false,
                            false
                    )
            );
        }
        resourcePacksInfoPacket.setForcedToAccept( false );
        resourcePacksInfoPacket.setForcingServerPacksEnabled( false );
        resourcePacksInfoPacket.setScriptingEnabled( false );
        player.getPlayerConnection().sendPacket( resourcePacksInfoPacket );
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/network/handler/LoginHandler.kt
package org.jukeboxmc.server.network.handler

import org.cloudburstmc.protocol.bedrock.data.DisconnectFailReason
import org.cloudburstmc.protocol.bedrock.packet.DisconnectPacket
import org.cloudburstmc.protocol.bedrock.packet.LoginPacket
import org.cloudburstmc.protocol.bedrock.packet.ServerToClientHandshakePacket
import org.cloudburstmc.protocol.bedrock.util.EncryptionUtils
import org.jukeboxmc.api.event.player.PlayerLoginEvent
import org.jukeboxmc.server.JukeboxServer
import org.jukeboxmc.server.network.LoginSequenceUtil
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.player.LoginData

class LoginHandler : PacketHandler<LoginPacket> {

    override fun handle(packet: LoginPacket, server: JukeboxServer, player: JukeboxPlayer) {
        player.setLoginData(LoginData().fromLoginPacket(packet))
        val playerLoginEvent = PlayerLoginEvent(player)

        if (!player.getLoginData()?.xboxAuthenticated!! && server.isOnlineMode()) {
            playerLoginEvent.setResult(PlayerLoginEvent.Result.XBOX_AUTHENTICATED)
            playerLoginEvent.setKickReason("You must be logged in with your xbox account.")
        }

        if (server.getOnlinePlayers().size >= server.getMaxPlayers()) {
            playerLoginEvent.setResult(PlayerLoginEvent.Result.SERVER_FULL)
            playerLoginEvent.setKickReason("Server is full.")
        }

        if (server.isWhitelist() && !server.getWhitelist().isWhitelisted(player.getName())) {
            playerLoginEvent.setResult(PlayerLoginEvent.Result.WHITELIST)
            playerLoginEvent.setKickReason("You are not on the whitelist")
        }

        server.getPluginManager().callEvent(playerLoginEvent)
        if (playerLoginEvent.getResult() != PlayerLoginEvent.Result.ALLOWED) {
            DisconnectPacket().apply {
                this.reason = DisconnectFailReason.KICKED
                this.kickMessage = playerLoginEvent.getKickReason()
                player.sendPacket(this)
            }
            return
        }

        if (server.isEncrypted()) {
            val token = EncryptionUtils.generateRandomToken()
            val serverToClientHandshakePacket = ServerToClientHandshakePacket()
            serverToClientHandshakePacket.jwt = EncryptionUtils.createHandshakeJwt(server.getKeyPair(), token)

            val session = player.getSession()
            val secretKey = EncryptionUtils.getSecretKey(server.getKeyPair().private, player.getLoginData()?.identityPublicKey!!, token)

            session.peer.channel.eventLoop().execute {
                session.sendPacketImmediately(serverToClientHandshakePacket)
                session.enableEncryption(secretKey)
            }
        } else {
            LoginSequenceUtil.processLoginSequence(server, player)
        }
    }
}